### PR TITLE
Sharing Block: automatically add block to templates in block themes

### PR DESCRIPTION
**Pull request by jeherve**
> [!IMPORTANT]
> This is an experiment to play with the Block Hooks API. It can only work with recent changes made to WP Core, and scheduled to ship in WordPress 6.5 next month.

## Proposed changes:

Under a few specific conditions, we want to automatically add a sharing buttons block (and 3 inner blocks, Facebook, X, and Mastodon) to the bottom of the post content.
Site owners can then go to Appearance > Editor if they want to remove or customize that block.

For now, we're trying to automatically add the block right after the post content block, in post and page templates (that includes templates like `page-wide`, `page-no-sidebar`, etc).

Here are the conditions:

- The site must be using a block theme.
- An option (`jetpack_sharing_buttons_auto_add`) must be set. Later on we can add UI to set that option under Jetpack > Settings > Social.
- The sharing module must be deactivated. We wouldn't want to automatically add sharing buttons to the post content twice.
- On WordPress.com Simple, the module is always active. There, we check if sharing posts are enabled on single posts and / or pages, since that matches the template where we'd like to automatically add the block.

**To do**

- [x] This doesn't quite work just yet, I haven't been able to figure out how to insert inner blocks. The site editor currently shows a validation error when viewing the single post template:
- [x] We should figure out how this should play with the sharedaddy module. Should we look at sharedaddy settings like we do right now and base our response on that? Or should we not look at it at all? In this case, when should auto-addition be triggered on WordPress.com Simple, where the sharing module is always active?

### Resources:

- https://make.wordpress.org/core/2023/10/15/introducing-block-hooks-for-dynamic-blocks/
- https://github.com/ockham/like-button
- https://core.trac.wordpress.org/ticket/59572
- https://core.trac.wordpress.org/ticket/60126

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

- p1707216098587689-slack-C065QF2R1B6

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

* Start by installing the development version of WordPress. Alternatively, [launch a quick test site via this link](https://jurassic.ninja/create?jetpack-beta&branches.jetpack=add/hook-sharing-block&wp-debug-log&wordpress-beta-tester&gutenberg).
* Go to Appearance > Themes and install a block-based theme.
* Go to Jetpack > Settings > Sharing and ensure that the Sharing buttons module is **off**.
* Via WP CLI, set the `jetpack_sharing_buttons_auto_add` option: `wp option add jetpack_sharing_buttons_auto_add true`
* Visit a single post on the frontend
    * Sharing buttons (Facebook, X, Mastodon) should be added below the post content.

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/sharing-button/sharing-button.php [hunk 1]
```diff
@@ -156,5 +161,135 @@ function sharing_process_requests() {
...
+			),
+		),
+		array(
+			'blockName' => INNER_BLOCK_NAME,
+			'attrs'     => array(
+				'service' => 'mastodon',
+				'label'   => esc_html__( 'Mastodon', 'jetpack' ),
+			),
+		),
+	);
```
Comment (ockham) [inline projects/plugins/jetpack/extensions/blocks/sharing-button/sharing-button.php:289]: The "parsed block" format is a bit weird; if you have `innerBlocks`, you also need to set `innerContent`. 

The latter is an array whose items are either HTML strings, or `null`, which tells WP/GB to substitute the `null` for the respective next block in the `innerBlocks` array.

That might sound a bit abstract; here's an example from https://github.com/WordPress/wordpress-develop/pull/5835 to illustrate:

```
// $hooked_block is something like array( 'blockName' => 'ockham/like-button', 'attrs' => array() );

// Wrap the Like Button block in a Group block.
return array(
	'blockName' => 'core/group',
	'attrs'     => $attrs,
	'innerBlocks' => array( $hooked_block ),
	'innerContent' => array(
		'<div class="wp-block-group">',
		null,
		'</div>'
	),
);
```

So at the very least, you'll need to set `innerContent => array( null, null, null )` (as you're trying to insert 3 inner blocks).

But you might need more than that: If you insert a `jetpack/sharing-buttons` manually in the editor and then view it in the code editor, does it include any HTML? If so, you'll have to include those HTML strings in `innerContent`.

Comment (jeherve) [reply projects/plugins/jetpack/extensions/blocks/sharing-button/sharing-button.php:289]: Thanks for the quick feedback! That was helpful, things are starting to look better with c0445f06fab289408d7b0cebe873bc4a2a7dd4ba :)

**Review by andrii-lysenko — COMMENTED** [review]
Thanks for working on this. This is a great POC and works as described.

I understand that it's not a ready solution yet, just wanted to point out what we also should consider in the future.
- We probably need to consider if Sharing Buttons already integrated as a part of Post template and not insert it if it is.
- We also won't need to add these if we add the block manually to the page.
- It would be great to see these buttons in editor as well, so they could be adjusted if needed.

**Review by andrii-lysenko — DISMISSED** [review]
Oh, I just got it completely wrong. We actually do change the template of the Post. I it's a great way to do this. 👍

**Comment by jeherve** [comment]
I think this is ready to go now. It won't appear for anyone since it relies on the `jetpack_sharing_buttons_auto_add` option that you can only set via CLI today, but it will allow us to start testing, and then to develop an interface for it in the Sharing settings screen in the Jetpack dashboard.

**Review by andrii-lysenko — APPROVED** [review]
I agree that it's safe to move forward as it won't actually show unless option is set tu `true`.
Let's merge it so we would be able to iterate on this.

**Pull request merged**